### PR TITLE
Patch May-July 2019

### PR DESCRIPTION
**Pull request by jongrumer**
##  CFF Patch: Improving RMCDHF and extending terms to J = 17/2 (l, m-)

In May @cffischer sent me a collection of updates, fixes and general improvements that I've put together into a single patch. The most important fixes are related to how the Lagrange multipliers are calculated in RMCDHF, and she has also extended the allowed terms on step to `J=17/2 (l, m-)`. I've also added some minor fixes to the rwfnmchfmcdf code and the ldlbl1 and 2 files in the transition codes which should remove some warnings during compilation.

Check the patch out and see what you think, review the code and let's get back to this in a week or so for merging. 

A description of the mods by Charlotte follows.
/Jon

### Patch summary (CFF)
'RMCDHF' is not robust in that accurate initial estimates are needed.
When algorithms fail, the messages are misleading, and in some instances 
the program simply stops. Several changes have been introduced that 
increase the robustness and the efficiency.

### Tests
The changes have been tested for a correlation study for Li 1s(2)2s(1) J=1/2 for wavefunctions from SDT excitations to nl=9l. 

### Libmod - increase J of allowed terms.
  
  1) `terms_C.f90`

Changes have been introduced that increase the maximum `J` to 17/2 
or `l` or `m-`. 
  
### Lib9290 - joining point evaluation has been improved

  1)  `setpot.f90`
 
This routine uses an algorithm that involves the orbital energy for
finding a join.  When the energy estimate is such that the criteria
for the join is not met, the program STOPS. The `STOP` has been 
replaced  by  `JP=NNNP/2`.  This sets the join to be the middle of the grid.

  2)  `start.90`

Remove the comment about accuracy because the problem is not related to
accuracy and is misleading.  

### RMCDHF

  1) `setlag.f90` and `setlagmpi.f90`

There was a logical  error in the section determining
Lagrange multilpliers.  When two subshells are
both varied the solution is not unique so the
desired solution is one where the lagrange
multiplier is zero.  But as soon as the orbitals
for one is fixed, a lagrange multiplier is needed.
Also, the order of the lagrange multipliers were 
not in the most efficient order. 

The calculation of Lagrange multipliers in the CSF 
procedure for heavy elements is the most time 
consuming task particularly the calculation of the 
exchange term. The order of computation has been 
modified so that POTL and XCH functions are called 
the minimum number of times.  In Francine (Fr) 
45 calls to exchange were reduced to 30 call.`When 
Correlation is included, XCH function includes 
contributions from all the interactions as well as 
exchange.

  2) `in.f90`     

When the algorithm for finding the outward boundary
condition did not find an end-point, because of poor initial
estimates an error message claimed the range was insufficient.
In the present code it is defined to be `NNNP-1`.  An
appropriate "end point" is determined in subsequent
iterations.  The range `NNNP=590` has been found adequate
for all test cases.

### Tools - increase dimension of the tools codes

Many of the routines in the `tool` library, do not use the GRASP 
parameters for determining dimensions and are subject to error.  In 
particular, "lines" of input data have character dimension of 100.  For 
heavy elements, this may not be sufficient and the program does not 
detect that information is lost.

The two corrected cases are,
   1) `rmixaccumulate.f90`
   2) `rmixextract.f90`

but there are probably more.

Actually, it is not a correction really since this just changes `100 => 200`. But 
there are several other cases that probably also are wrong.  Unless 
someone else updates the online GRASP2018 code, Charlotte said she will do that as soon, possibly this summer. `rmixextract.f90` also had another error that she fixed.

## Additional fixes (JG)
  1) `tools/rwfnmchfmcdf.f90`

Bugfix: initialization loop of arrays were done in an old fashioned style, but also from element 0 which gave a warning during compilation. This is now changed to simple `A = 0` statements.

  2) `rtransition(_mpi)/ldlbl1(2).f90`

Fixed legacy-style read statements.

### CODE REVIEW

**Comment by cffischer** [comment]
I have reviewed the changes --- all look good.

**Comment by cffischer** [comment]
OK

File: src/tool/rwfnmchfmcdf.f90 [hunk 1]
```diff
@@ -35,18 +35,22 @@ program ff2gr
...
 
 
-      do 2 i=0,230
-         pg(i)=0.d0
-         qg(i)=0.d0
-2     continue
+!      do 2 i=0,230
+!         pg(i)=0.d0
+!         qg(i)=0.d0
+!2     continue
```
[the remark this answers is not in the document]
Comment (jongrumer) [reply src/tool/rwfnmchfmcdf.f90]: Yes of course, my mistake. Thanks @mortenpi!

File: src/appl/rmcdhf90/in.f90 [hunk 2]
```diff
@@ -143,7 +143,8 @@ SUBROUTINE IN(IORB, JP, P, Q, MTP)
 !
       TLAST = TTHIS
       TTHIS = ABS(XS(I)/TI(I))
-      IF (TTHIS + TLAST <= EPS) THEN
+!cff  Stop if J=NNNP
```
[the remark this answers is not in the document]
Comment (cffischer) [reply src/appl/rmcdhf90/in.f90]: It is not a comment. I commented out the statement.  Probably better to delete the statement.

File: src/appl/rmcdhf90/setlag.f90 [hunk 3]
```diff
@@ -85,65 +86,30 @@ SUBROUTINE SETLAG(EOL)
...
                FULLJ = ABS ( UCF(LJ)-DBLE (NKJ(LJ)+1) ) .LT. EPS
-               IF ( (NAK(LJ) .EQ. NAKLI) .AND.            &
-                     (.NOT. (FIXLI .AND. FIXLJ)) .AND.    &
-                     (.NOT. (FULLI .AND. FULLJ)) ) THEN
+            DO LIraw = 1, LJraw-1
+               LI = iorder(LIraw)
+!cff        March, 2019
+               VLI = .NOT. LFIX(LI)
+               FULLI = ABS ( UCF(LI)-DBLE (NKJ(LI)+1) ) .LT. EPS
+!cff         March, 2019
```
[the remark this answers is not in the document]
Comment (cffischer) [reply src/appl/rmcdhf90/setlag.f90]: I agree.   This is what we used to do before git.

File: src/appl/rmcdhf90/setlag.f90 [hunk 4]
```diff
@@ -258,20 +225,36 @@ SUBROUTINE SETLAG(EOL)
...
+!              TA(I) = UCFM*(PF(I,J)*XQM(I)-QF(I,J)*XPM(I)) +           &
+!                      UCFJ*(PF(I,M)*XQJ(I)-QF(I,M)*XPJ(I)) 
+!              TA(I) = RPOR(I)*(TA(I)*C +                               &
+!                      (PF(I,J)*PF(I,M)+QF(I,J)*QF(I,M))*               &
+!                      (UCFM*YPM(I) + UCFJ*YPJ(I)))
+!           END DO 
+!           CALL QUAD (RESULT) 
+!
+!           RIJM = RINTI(J,M,1)                 
+!           ECV(LI) = (RESULT - (UCFM+UCFJ)*RIJM)/2
```
[the remark this answers is not in the document]
Comment (cffischer) [reply src/appl/rmcdhf90/setlag.f90]: This is commented out code. There is a lot of this in GRASP.

File: src/appl/rmcdhf90/setlag.f90 (hunk 3 above)
[the remark this answers is not in the document]
Comment (jongrumer) [reply src/appl/rmcdhf90/setlag.f90]: Great! I'll remove them then!

File: src/appl/rmcdhf90/in.f90 (hunk 2 above)
[the remark this answers is not in the document]
Comment (jongrumer) [reply src/appl/rmcdhf90/in.f90]: Ok!

File: src/appl/rmcdhf90/setlag.f90 (hunk 4 above)
[the remark this answers is not in the document]
Comment (jongrumer) [reply src/appl/rmcdhf90/setlag.f90]: So, what do you think @cffischer - should I remove it? The less "alternative" code we have, the easier it is to understand it.

[the remark this answers is not in the document]
Comment (cffischer) [reply src/appl/rmcdhf90/setlag.f90]: Yes.   This should then be a policy decision that "commented out code" should be deleted.  This is not something we did before GitHub.

[the remark this answers is not in the document]
Comment (jongrumer) [reply src/appl/rmcdhf90/setlag.f90]: Great - we should get started on a coding policy at some point. I'll put it on my to do list :)

**Comment by jongrumer** [comment]
Ok, I think we're good to go. Will go on vacation on Tuesday, so will probably try to merge this into master on Monday latest if there are no further objections.

File: src/appl/rmcdhf90_mpi/setlagmpi.f90 [hunk 5]
```diff
@@ -85,61 +86,28 @@ SUBROUTINE SETLAGmpi(EOL)
...
                FULLJ = ABS ( UCF(LJ)-DBLE (NKJ(LJ)+1) ) .LT. EPS
-               IF ( (NAK(LJ) .EQ. NAKLI) .AND.            &
-                     (.NOT. (FIXLI .AND. FIXLJ)) .AND.    &
-                     (.NOT. (FULLI .AND. FULLJ)) ) THEN
+            DO LIraw = 1, LJraw-1
+               LI = iorder(LIraw)
+!cff           March, 2019 
+               VLI = .NOT. LFIX(LI)
+               FULLI = ABS ( UCF(LI)-DBLE (NKJ(LI)+1) ) .LT. EPS
+!cff           March, 2019
```
Comment (mortenpi) [inline src/appl/rmcdhf90_mpi/setlagmpi.f90]:
```suggestion
```

File: src/appl/rmcdhf90_mpi/setlagmpi.f90 [hunk 6]
```diff
@@ -85,61 +86,28 @@ SUBROUTINE SETLAGmpi(EOL)
...
+            NAKLJ = NAK(LJ)
+!cff        March, 2019 
+            VLJ = .NOT. LFIX(LJ)
                FULLJ = ABS ( UCF(LJ)-DBLE (NKJ(LJ)+1) ) .LT. EPS
-               IF ( (NAK(LJ) .EQ. NAKLI) .AND.            &
-                     (.NOT. (FIXLI .AND. FIXLJ)) .AND.    &
-                     (.NOT. (FULLI .AND. FULLJ)) ) THEN
+            DO LIraw = 1, LJraw-1
+               LI = iorder(LIraw)
+!cff           March, 2019
```
Comment (mortenpi) [inline src/appl/rmcdhf90_mpi/setlagmpi.f90]:
```suggestion
```

File: src/appl/rmcdhf90_mpi/setlagmpi.f90 [hunk 7]
```diff
@@ -85,61 +86,28 @@ SUBROUTINE SETLAGmpi(EOL)
...
-            LIP1 = MAX (NCORE, LIraw) + 1
-            NAKLI = NAK(LI)
-            FIXLI = LFIX(LI)
-            FULLI = ABS ( UCF(LI)-DBLE (NKJ(LI)+1) ) .LT. EPS
-            DO LJraw = LIP1, NW
+         DO LJraw = NCORE+1, NW
                LJ = iorder(LJraw)
-               FIXLJ = LFIX(LJ)
+            NAKLJ = NAK(LJ)
+!cff        March, 2019
```
Comment (mortenpi) [inline src/appl/rmcdhf90_mpi/setlagmpi.f90]:
```suggestion
```

File: src/appl/rmcdhf90_mpi/setlagmpi.f90 [hunk 8]
```diff
@@ -56,11 +56,12 @@ SUBROUTINE SETLAGmpi(EOL)
 !-----------------------------------------------
 !   L o c a l   V a r i a b l e s
 !-----------------------------------------------
-      INTEGER :: ITWICE, LIRAW, LI, LIP1, NAKLI, LJRAW, LJ, IECCLI, L1, L2, &
+      INTEGER :: ITWICE, LIRAW, LI, LIP1, NAKLJ, LJRAW, LJ, IECCLI, L1, L2, &
          JLAST, MLAST, M, J, I
       REAL(DOUBLE), DIMENSION(NNNP) :: YPJ, YPM, XPJ, XPM, XQJ, XQM
       REAL(DOUBLE) :: EPS,UCFJ,UCFM,RESULT,RIJM,QDIF,OBQDIF,OBQSUM,TMP
-      LOGICAL :: FIRST, FIXLI, FIXLJ, FULLI, FULLJ
+!cff  March 2019
```
Comment (mortenpi) [inline src/appl/rmcdhf90_mpi/setlagmpi.f90]:
```suggestion
```

**Comment by jongrumer** [comment]
Aight, removed the final timestamps suggested by @mortenpi. Will merge now then!

**Comment by jongrumer** [comment]
Done! Thanks to @cffischer for the update and @mortenpi for the reviews. Let's hope everything works as expected now then.

**Pull request merged**